File: app/src/main/java/com/withever/blind/base/recyclerview/LinearLayoutSmoothScrollerManager.java
package com.withever.blind.base.recyclerview;

import android.content.Context;
import android.graphics.PointF;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.LinearSmoothScroller;
import androidx.recyclerview.widget.RecyclerView;

public class LinearLayoutSmoothScrollerManager extends LinearLayoutManager {

    private static final float MILLISECONDS_PER_INCH = 50f;
    private Context context;

    public LinearLayoutSmoothScrollerManager(Context context) {
        super(context);
        this.context = context;
    }

    public LinearLayoutSmoothScrollerManager(Context context, int orientation, boolean reverseLayout) {
        super(context, orientation, reverseLayout);
        this.context = context;
    }

    public LinearLayoutSmoothScrollerManager(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        this.context = context;
    }

    @Override
    public void smoothScrollToPosition(RecyclerView recyclerView, RecyclerView.State state, final int position) {

        LinearSmoothScroller smoothScroller = new LinearSmoothScroller(context) {
            @Override
            public PointF computeScrollVectorForPosition(int targetPosition) {
                return LinearLayoutSmoothScrollerManager.this.computeScrollVectorForPosition(targetPosition);
            }

            @Override
            protected float calculateSpeedPerPixel
                    (DisplayMetrics displayMetrics) {
                return MILLISECONDS_PER_INCH / displayMetrics.densityDpi;
            }
        };

        smoothScroller.setTargetPosition(position);
        startSmoothScroll(smoothScroller);
    }

//    @Override
//    public void smoothScrollToPosition(RecyclerView recyclerView, RecyclerView.State state, int position) {
//        final LinearSmoothScroller linearSmoothScroller = new LinearSmoothScroller(recyclerView.getContext()) {
//            private static final float MILLISECONDS_PER_INCH = 100f;
//
//            @Override
//            public PointF computeScrollVectorForPosition(int targetPosition) {
//                return LinearLayoutSmoothScrollerManager.this.computeScrollVectorForPosition(targetPosition);
//            }
//
//            @Override
//            protected float calculateSpeedPerPixel(DisplayMetrics displayMetrics) {
//                return MILLISECONDS_PER_INCH / displayMetrics.densityDpi;
//            }
//        };
//        linearSmoothScroller.setTargetPosition(position);
//        startSmoothScroll(linearSmoothScroller);
//    }


//    @Override
//    public void smoothScrollToPosition(RecyclerView recyclerView, RecyclerView.State state, int position) {
//        final LinearSmoothScroller linearSmoothScroller =
//                new LinearSmoothScroller(recyclerView.getContext()) {
//                    private static final float MILLISECONDS_PER_INCH = 15f;
//
//                    @Override
//                    public PointF computeScrollVectorForPosition(int targetPosition) {
//                        return super.computeScrollVectorForPosition(targetPosition);
//                    }
//
//                    @Override
//                    protected float calculateSpeedPerPixel(DisplayMetrics displayMetrics) {
//                        return MILLISECONDS_PER_INCH / displayMetrics.densityDpi;
//                    }
//                };
//        linearSmoothScroller.setTargetPosition(position);
//        startSmoothScroll(linearSmoothScroller);
//    }

}

File: app/src/main/java/com/withever/blind/base/recyclerview/LinearLayoutSmoothScrollerManager.kt
package com.withever.blind.base.recyclerview

import android.content.Context
import android.graphics.PointF
import android.util.AttributeSet
import android.util.DisplayMetrics
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearSmoothScroller
import androidx.recyclerview.widget.RecyclerView

class LinearLayoutSmoothScrollerManager : LinearLayoutManager {

    private var context: Context

    constructor(context: Context) : super(context){
       this.context = context
    }

    constructor(context: Context, orientation: Int, reverseLayout: Boolean)
            : super(context, orientation, reverseLayout){
        this.context = context
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes){
        this.context = context
    }

    override fun smoothScrollToPosition(
        recyclerView: RecyclerView?,
        state: RecyclerView.State?,
        position: Int
    ) {
        val smoothScroller: LinearSmoothScroller = object :LinearSmoothScroller(context){
            override fun computeScrollVectorForPosition(targetPosition: Int): PointF? {
                return this@LinearLayoutSmoothScrollerManager.computeScrollVectorForPosition(targetPosition)
            }

            override fun calculateSpeedPerPixel(displayMetrics: DisplayMetrics): Float {
                return MILLISECONDS_PER_INCH / displayMetrics.densityDpi
            }
        }

        smoothScroller.targetPosition = position
        startSmoothScroll(smoothScroller)
    }

    companion object{
        private const val MILLISECONDS_PER_INCH: Float = 50f
    }
}
